compiler/rustc_passes/src/liveness.rs

```diff
@@ -803,38 +827,67 @@ impl<'a, 'tcx> Liveness<'a, 'tcx> {
                 // initialization, which is mildly more complex than checking
                 // once at the func header but otherwise equivalent.
 
-                let succ = self.propagate_through_opt_expr(local.init, succ);
-                self.define_bindings_in_pat(&local.pat, succ)
+                let (succ, succ_div) =
+                    self.propagate_through_opt_expr(local.init, succ, warn_about_unreachable);
+                (self.define_bindings_in_pat(&local.pat, succ), succ_div)
             }
-            hir::StmtKind::Item(..) => succ,
+            hir::StmtKind::Item(..) => (succ, false),
             hir::StmtKind::Expr(ref expr) | hir::StmtKind::Semi(ref expr) => {
-                self.propagate_through_expr(&expr, succ)
+                self.propagate_through_expr(&expr, succ, warn_about_unreachable)
             }
         }
     }
 
-    fn propagate_through_exprs(&mut self, exprs: &[Expr<'_>], succ: LiveNode) -> LiveNode {
-        exprs.iter().rev().fold(succ, |succ, expr| self.propagate_through_expr(&expr, succ))
+    fn propagate_through_exprs(
+        &mut self,
+        exprs: &[Expr<'_>],
+        succ: LiveNode,
+        warn_about_unreachable: bool,
```

Comment:
Can we factor out a `enum WarnUnreachable { Yes, No }` for this? That way it's clear at the call site what is getting passed around.

Reply:
Good idea; I've added two new enums `WarnUnreachable` and `Diverges`.